shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,773 @@
+
+function backpropagateVertex(context, vertex, permissionSet, viewInfo) {
+  // Computes the flow of the permissions in `permissionSet` from the grain owner to `vertex`.
+  // Returns a map whose values are MembranedPermissionSets and whose keys are the Clause.hash()
+  // of the corresponding membrane.
+  check(context, Context);
+  check(vertex, vertexPattern);
+  check(permissionSet, PermissionSet);
+
+  const grainId = vertex.token ? vertex.token.grainId : vertex.grain._id;
+  const grain = context.grains[grainId];
+  const ownerIdentityIds = context.userIdentityIds[grain.userId];
+
+  // A vertex ID is of the form "i:<identityId>" or "t:<tokenId>".
+  const destinationId = vertex.token ? ("t:" + vertex.token._id) : ("i:" + vertex.grain.identityId);
+  const destinationPermissions = {};
+  {
+    const clause = new Clause();
+    destinationPermissions[clause.hash()] = new MembranedPermissionSet(permissionSet, clause);
   }
 
-  var openerIdentityId = edges.terminalEdge.sharer;
-  var edgesByRecipient = edges.edgesByRecipient;
-  var owner = "OwningAccount";
+  const permissionsMap = {};
+  // Map<vertexId, Map<membraneHash, MembranedPermissionSet>>.
+  // May from vertex ID to permissions that we've already shown flow from that vertex to our
+  // destination vertex.
 
-  var permissionsMap = {};
-  // Keeps track of the permissions that the opener receives from each user. The final result of
-  // our computation will be stored in permissionsMap[owner].
+  const vertexStack = []; // Vertex IDs that we need to explore.
 
-  var userStack = [openerIdentityId];
+  permissionsMap[destinationId] = destinationPermissions;
+  vertexStack.push(destinationId);
 
-  var openerAttenuation = roleAssignmentPermissions({allAccess: null}, viewInfo);
-  edges.terminalEdge.roleAssignments.forEach(function(roleAssignment) {
-    openerAttenuation.intersect(roleAssignmentPermissions(roleAssignment, viewInfo));
-  });
-  permissionsMap[openerIdentityId] = openerAttenuation;
-
-  while (userStack.length > 0) {
-    var recipient = userStack.pop();
-    if (edgesByRecipient[recipient]) {
-      edgesByRecipient[recipient].forEach(function (inEdge) {
-        var sharer = inEdge.sharer;
-        var needToPush = false;
-        if (!permissionsMap[sharer]) {
-          permissionsMap[sharer] = new PermissionSet();
-          needToPush = true;
-        }
+  while (vertexStack.length > 0) {
+    const vertexId = vertexStack.pop();
+    let incomingEdges = [];
 
-        var newPermissions = roleAssignmentPermissions({allAccess: null}, viewInfo);
-        inEdge.roleAssignments.forEach(function(roleAssignment) {
-          newPermissions.intersect(roleAssignmentPermissions(roleAssignment, viewInfo));
-        });
-        newPermissions.intersect(permissionsMap[recipient]);
+    function tokenToEdge(token) {
+      let sharerId = "i:" + token.identityId;
+      if (token.parentToken) {
```

Comment:
I feel like a ternary operator would be more readable, and would also avoid computing a string concatenation that just gets thrown away in the case that there's a parent token.